openquake/hazardlib/contexts.py

```diff
@@ -1094,20 +1116,55 @@ def get_cs_contrib(self, ctx, imti, imls):
             # 5: third site
             # i.e. idxs = [0, 3], [1, 4], [2, 5] for sites 0, 1, 2
             slc = slice(n, N * U, N)  # U indices
+
+            # For every GMM
             for g in range(G):
                 mu = mean_stds[0, g, :, slc]  # shape (M, U)
                 sig = mean_stds[1, g, :, slc]  # shape (M, U)
+
                 c = out[self.start + g]['_c']
                 s = out[self.start + g]['_s']
+
+                # For each IML
                 for p in range(P):
-                    eps = (imls[p] - mu[imti]) / sig[imti]  # shape U
-                    poes = _truncnorm_sf(self.truncation_level, eps)  # shape U
+
+                    # Calculate the contribution of each rupture to the total
+                    # probability of occurrence for the reference IMT. Both
+                    # `eps` and `poes` have shape U
+                    eps = (numpy.log(imls[p]) - mu[imti]) / sig[imti]
+                    poes = _truncnorm_sf(self.truncation_level, eps)
+
+                    # Converting to rates and dividing by the rate of
+                    # exceedance of the reference IMT and level
                     ws = -numpy.log(
                         (1. - probs[slc]) ** poes) / self.investigation_time
+
+                    # Normalizing by the AfE for the investigated IMT and level
+                    ws /= -numpy.log(1. - cs_poes[p])
+
+                    # If sum of ws is close to 1 we normalize
+                    if numpy.abs(ws.sum() - 1.0) < 1e-2:
+                        ws /= ws.sum()
+
+                    # Normalising weights
                     s[n, p] = ws.sum()  # weights not summing up to 1
+
+                    # For each intensity measure type
                     for m in m_range:
-                        c[m, n, 0, p] = ws @ (mu[m] + rho[m] * eps * sig[m])
-                        c[m, n, 1, p] = ws @ (sig[m]**2 * (1. - rho[m]**2))
+
+                        # Equation 14 in Lin et al. (2013)
+                        term1 = (mu[m] + rho[m] * eps * sig[m])
+                        c[m, n, MEA, p] = ws @ term1
+
+                        # This is executed only if we already have the final CS
+                        if len(cs) > 0:
+
+                            # Equation 15 in Lin et al. (2013)
+                            term2 = sig[m] * (1. - rho[m]**2)**0.5
+                            term3 = (term1 - cs[0]['_c'][m, n, 0, p])
+                            term3 = 0
```

Comment:
This is certainly an error. First you define `term3` and then you reset it to zero. What should it be?